boot/bootutil/src/swap_move.c

```diff
@@ -0,0 +1,506 @@
+                 "image_ok=0x%x",                                   \
+                 (area),                                            \
+                 ((state)->magic == BOOT_MAGIC_GOOD ? "good" :      \
+                  (state)->magic == BOOT_MAGIC_UNSET ? "unset" :    \
+                  "bad"),                                           \
+                 (state)->swap_type,                                \
+                 (state)->copy_done,                                \
+                 (state)->image_ok)
+
+int
+boot_status_source(struct boot_loader_state *state)
+{
+    struct boot_swap_state state_primary_slot;
+    int rc;
+    uint8_t source;
+    uint8_t image_index;
+
+#if (BOOT_IMAGE_NUMBER == 1)
+    (void)state;
+#endif
+
+    image_index = BOOT_CURR_IMG(state);
+
+    rc = boot_read_swap_state_by_id(FLASH_AREA_IMAGE_PRIMARY(image_index),
+            &state_primary_slot);
+    assert(rc == 0);
+
+    BOOT_LOG_SWAP_STATE("Primary image", &state_primary_slot);
+
+    if (state_primary_slot.magic == BOOT_MAGIC_GOOD &&
+            state_primary_slot.copy_done == BOOT_FLAG_UNSET) {
+
+        source = BOOT_STATUS_SOURCE_PRIMARY_SLOT;
+
+        BOOT_LOG_INF("Boot source: primary slot");
+        return source;
+    }
+
+    BOOT_LOG_INF("Boot source: none");
+    return BOOT_STATUS_SOURCE_NONE;
+}
+
+/*
+ * "Moves" the sector located at idx - 1 to idx.
+ */
+static void
+boot_move_sector_up(int idx, uint32_t sz, struct boot_loader_state *state,
+        struct boot_status *bs)
+{
+    const struct flash_area *fap_primary_slot;
+    const struct flash_area *fap_secondary_slot;
+    uint32_t new_off;
+    uint32_t old_off;
+    uint8_t image_index;
+    int rc;
+
+    /*
+     * FIXME: assuming sectors of size == sz, a single off variable
+     * would be enough
+     */
+
+    /* Calculate offset from start of image area. */
+    new_off = boot_img_sector_off(state, BOOT_SECONDARY_SLOT, idx);
+    old_off = boot_img_sector_off(state, BOOT_SECONDARY_SLOT, idx - 1);
+
+    image_index = BOOT_CURR_IMG(state);
+
+    rc = flash_area_open(FLASH_AREA_IMAGE_PRIMARY(image_index),
+            &fap_primary_slot);
+    assert (rc == 0);
+
+    rc = flash_area_open(FLASH_AREA_IMAGE_SECONDARY(image_index),
+            &fap_secondary_slot);
+    assert (rc == 0);
+
+    if (bs->idx == BOOT_STATUS_IDX_0) {
+        rc = boot_erase_trailer_sectors(state, fap_primary_slot);
+        assert(rc == 0);
+
+        rc = boot_status_init(state, fap_primary_slot, bs);
+        assert(rc == 0);
+
+        rc = boot_erase_trailer_sectors(state, fap_secondary_slot);
+        assert(rc == 0);
+    }
+
+    rc = boot_erase_region(fap_secondary_slot, new_off, sz);
+    assert(rc == 0);
+
+    rc = boot_copy_region(state, fap_secondary_slot, fap_secondary_slot,
+                          old_off, new_off, sz);
+    assert(rc == 0);
+
+    rc = boot_write_status(state, bs);
+
+    bs->idx++;
```

Comment:
I would increment `bs->idx` after return from call to `this`.
just a matter of taste, but I thing it will make algorithm more legible from higher level.

Reply:
I disagree; this function shares responsibilities with `boot_swap_sectors`, and they both modify the state internally. Would be weird if one modified the state while the other not.

Reply:
ok